### ISSUE TITLE
Repo approach 

### ISSUE THREAD

**Issue by kzisme** [kzisme/tpython#1]
So, @whit537 Do you think it would be beneficial to go through the first few chapters?  

Also, would a folder for each chapter be necessary?  

What do you like to do when approaching something like this?  I just want to get something out of it (maybe someone could look at it when they are beginning as well! (although I'm just using this to re-learn some stuff) :smile_cat:

**Comment by chadwhitacre** [comment]
> What do you like to do when approaching something like this?

I find it helpful to put together a schedule. How many weeks do you want to spend on this? How much time per week? Based on that, split the book up into chunks. Presumably you want to do the reading _and_ the exercises, ya? How about posting the exercises to this repo each Wednesday and I can code-review them with you?

http://www.greenteapress.com/thinkpython/html/

**Comment by kzisme** [comment]
@whit537 Yeah - more the exercises than anything - reading along with it.  The only thing I'm not terribly interested in is Swampy.  

Sounds good to me!  Is it worth making a new folder for each week or just creating a new .py file for each week?

**Comment by chadwhitacre** [comment]
@kzisme Up to you, new folder probably makes sense. Basically, it's your repo ... I'm happy to follow along and give you feedback and answer any questions that come up! :-)

**Comment by kzisme** [comment]
@whit537 I just want this to be a refresher - and to be more comfortable with Python than I was before.  So, I'm open to any suggestions/criticisms on code/repo/anything

**Comment by kzisme** [comment]
@whit537 I created the chapters (some of which I don't plan to look at or work on)

Thus far I've found out that Git is unable to find/add empty directories - and I'm not sure why the directories are out of order when listed...that bothers me

**Comment by chadwhitacre** [comment]
> I'm not sure why the directories are out of order when listed...that bothers me

Try `chapter01` instead of `chapter1`, etc.—with a `0`.

**Comment by chadwhitacre** [comment]
> Thus far I've found out that Git is unable to find/add empty directories

You're learning already! :-)

Another idea: if you name each file `README.md` instead of `description.md` then GitHub will display the content as the index for that directory.

**Comment by kzisme** [comment]
Is there a way to automate this rather than navigating into each directory
and then creating the files? (20+ times)

**Comment by chadwhitacre** [comment]
Yes! You should be able to use a bash for loop.

**Comment by chadwhitacre** [comment]
Something like this? (untested)

``` bash
for x in $( ls ); do git mv $x/description.md $x/README.md done
```

http://tldp.org/HOWTO/Bash-Prog-Intro-HOWTO-7.html

**Comment by kzisme** [comment]
Tried that - kept getting a syntax error (I haven't written much bash)

So....

I just used:  

```
find . -name .DS_Store -print0 | xargs -0 git rm -f --ignore-unmatch
```

To remove all the irritating .DS_store files I created by removing the description.md files

Currently there is just placeholder text from the top level directory readme - I'll change those as I go along.

I plan to:
1. Make notes in the markdown - per chapter 
2. Include .py files with the required exercises 

I suppose you can either review them at your leisure - or every Wednesday I want to try to do a little each day , but other things may get in the way.

**Comment by chadwhitacre** [comment]
Cool. I'll keep an eye on the repo, and we'll see how it goes! :-)

**Comment by kzisme** [comment]
I appreciate any feedback you have - and for your time @whit537 !m

**Comment by kzisme** [comment]
@whit537 I've completed up to chapter 7 - most of the exercises I just did the smaller ones as the others seemed boring (to me at least?)

If you have suggestions on exercises please let me know - even if it's a small project using the different chapters and such.  I've already taken a Python course that used this book too (or at least we did most of the exercises from the book)

Other questions for you:
1. What is your development environment like?  I use vim and that's about it right now.  I'm also running on Python 2.7.10 (I need to update to 3x but I haven't looked into that yet)
2. I haven't reached the specific chapter yet but...what is the difference between an iterator and an iterable?

I was referring to [this](http://stackoverflow.com/questions/9884132/what-exactly-are-pythons-iterator-iterable-and-iteration-protocols) post to figure it out but it still seemed confusing.  

I may come up with more questions later as well...

Feel free to look over stuff at your leisure - I appreciate the feedback as always :dancer:

**Comment by kzisme** [comment]
@whit537 May try to work through chapter 8 tonight before I go to sleep. :dancer:

**Comment by chadwhitacre** [comment]
> What is your development environment like? I use vim and that's about it right now. I'm also running on Python 2.7.10 (I need to update to 3x but I haven't looked into that yet)

I use Vim with the [pyflakes](https://github.com/kevinw/pyflakes-vim) plugin (/me discovers it's deprecated!). I upgraded my global Python to 3.5 just a couple months ago. Before that I was on 2.7. Of course for Gratipay and other projects I still use 2.7. Python 3 is here! :dancer: 

> I haven't reached the specific chapter yet but...what is the difference between an iterator and an iterable?

I did some playing around based on the Stack Overflow answer you linked, Looks like an **iterable** is a class with an **iter** method, and an **iterator** is what you get when you _call_ that method.

``` python
class Iterable:
    one = 1
    two = 2
    three = 3

    def __iter__(self):
        yield self.one
        yield self.two
        yield self.three
```

``` python
$ python
Python 3.5.0 (default, Sep 25 2015, 10:18:27) 
[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.72)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import itera
>>> i = itera.Iterable()
>>> i.__next__
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'Iterable' object has no attribute '__next__'
>>> i.__iter__
<bound method Iterable.__iter__ of <itera.Iterable object at 0x10a607048>>
>>> for j in i: print(j)
... 
1
2
3
>>> iterator = i.__iter__()
>>> iterator
<generator object Iterable.__iter__ at 0x10a600468>
>>> iterator.__next__
<method-wrapper '__next__' of generator object at 0x10a600468>
>>>
```

**Comment by chadwhitacre** [comment]
Figured [it](https://github.com/kzisme/tpython/issues/1#issuecomment-161094648) out!

``` bash
$ for foo in `ls 2015*pdf`; do git mv $foo 2015/; done
```

We were missing the semicolon before `done`. :)

**Comment by kzisme** [comment]
Awesome!  Sorry for slacking on this - busy during the holidays as usual.  I appreciate you still checking in on this/me.

I got two new Python books for Christmas one is relating to Data Structures and the other is best practices.  The one topic I'm having a hard time understanding is Big O notation and how to tell the notation of a code fragment (I have an idea on most but not a 100% way to determine it)

Any suggestions on material regarding that @whit537

**Comment by chadwhitacre** [comment]
@kzisme I studied up on Big O for a while when I was trying to get a job at Google. I think it's mostly a matter of finding and working on problems where it actually matters. I've never really had to use Big O that much in the real world, but that could be because of the kind of stuff I've worked on.

**Comment by kzisme** [comment]
@whit537 Sorry for not being active much!  Been busy lately as well as getting a part time dev job between school times :smile_cat: 

I also got my own website (finally) at a decent state www.kzisme.com do you have any critiques of it?  (aside from lack of content)

Edit: Another thing I've been working on in my spare time with a friend is a [chrome extension](https://github.com/kzisme/HnTagger) If you have a moment can you review it?  It's still not great but it was just refactored and we changed the way some stuff was working (Ideally I can apply this to knowledge to the Gratipay extension)
